File: Exercise/Hard/08. Roman Numeral Validator and Converter/Code/Roman Numeral Validator and Converter.c
#include <stdio.h>
#include <stdbool.h>

int romanToInt(char *s) {
    int romanValues[26];
    romanValues['I' - 'A'] = 1;
    romanValues['V' - 'A'] = 5;
    romanValues['X' - 'A'] = 10;
    romanValues['L' - 'A'] = 50;
    romanValues['C' - 'A'] = 100;
    romanValues['D' - 'A'] = 500;
    romanValues['M' - 'A'] = 1000;

    int result = 0, prevValue = 0;

    for (int i = 0; s[i] != '\0'; i++) {
        int currentValue = romanValues[s[i] - 'A'];

        if (currentValue > prevValue) {
            result += (currentValue - 2 * prevValue);
        } else {
            result += currentValue;
        }

        prevValue = currentValue;
    }

    return result;
}

bool isValidRoman(char *s) {
    for (; *s; s++) {
        if (*s != 'I' && *s != 'V' && *s != 'X' && *s != 'L' &&
            *s != 'C' && *s != 'D' && *s != 'M') {
            return false;
        }
    }
    return true;
}

int main() {
    char roman[100];
    scanf("%s", roman);

    if (!isValidRoman(roman)) {
        printf("Error: Invalid Roman numeral\n");
    } else {
        int arabic = romanToInt(roman);
        printf("%d\n", arabic);
    }

    return 0;
}

File: Exercise/Hard/08. Roman Numeral Validator and Converter/Code/Roman Numeral Validator and Converter.cs
using System;

class Program {
    static int RomanToInt(string s) {
        var romanValues = new System.Collections.Generic.Dictionary<char, int> {
            {'I', 1}, {'V', 5}, {'X', 10}, {'L', 50},
            {'C', 100}, {'D', 500}, {'M', 1000}
        };

        int result = 0;
        int prevValue = 0;

        for (int i = s.Length - 1; i >= 0; i--) {
            int currentValue = romanValues[s[i]];

            if (currentValue >= prevValue) {
                result += currentValue;
            }
            else {
                result -= currentValue;
            }

            prevValue = currentValue;
        }

        return result;
    }

    static bool IsValidRoman(string s) {
        foreach (char c in s) {
            if (c != 'I' && c != 'V' && c != 'X' && c != 'L' &&
                c != 'C' && c != 'D' && c != 'M')
            {
                return false;
            }
        }
        return true;
    }

    static void Main() {
        string roman = Console.ReadLine();

        if (!IsValidRoman(roman)) {
            Console.WriteLine("Error: Invalid Roman numeral");
        }
        else {
            int arabic = RomanToInt(roman);
            Console.WriteLine(arabic);
        }
    }
}



